src/condor_starter.V6.1/filter.cpp

```diff
@@ -114,67 +98,42 @@ filterPluginResults( const classad::ClassAd & ad ) {
+        // Reserved for future expansion.
+        { "TransferData",           Value::LIST_VALUE },
     };
 
+    classad::ClassAd * filteredAd = filterClassAd( ad, LegalAttributes );
 
-    // The ClassAd iterator isn't sorted, which set_intersection() requires.
-    classad::References attributes;
-    sGetAdAttrs( attributes, ad );
-
-    std::vector<std::string> intersection;
-#if defined(CXX_TWENTY_IS_REAL_TO_ME)
-    std::ranges::set_intersection(
-        attributes,
-        LegalAttributes | std::ranges::views::keys,
-        std::back_inserter( intersection ),
-        CaseIgnLTStr()
-    );
-#else
-    for( const auto & attribute : attributes ) {
-        if( LegalAttributes.count(attribute) != 0 ) {
-            intersection.push_back( attribute );
+    // Additionally, filter the TransferErrorData attribute.
+    ExprTree * e = filteredAd->Lookup( "TransferErrorData" );
+    if( e != NULL ) {
+        if( e->GetKind() != ExprTree::EXPR_LIST_NODE ) {
+            EXCEPT("TransferErrorData filtered in but not a list");
         }
-    }
-#endif
 
+        ExprList * list = dynamic_cast<ExprList *>(e);
+        if( list == NULL ) {
+            EXCEPT("Failed to cast EXPR_LIST_NODE to ExprList");
+        }
 
-    if( intersection.empty() ) { return NULL; }
 
-    classad::ClassAd * filteredAd = new classad::ClassAd();
-    for( const auto & attribute : intersection ) {
-        classad::Value v;
-        ExprTree * e = ad.Lookup( attribute );
-        if( e->Evaluate(v) ) {
-                if( v.GetType() == LegalAttributes[attribute] ) {
-                    if( attribute == "TransferErrorData" ) {
-                        ClassAd * tedAd = NULL;
-                        ASSERT(v.IsClassAdValue(tedAd));
-                        ClassAd * filteredTEDAd = filterTransferErrorData(* tedAd);
-                        if( filteredTEDAd != NULL ) {
-                            filteredAd->Insert( attribute, filteredTEDAd );
-                        }
-                    } else {
-                        filteredAd->Insert( attribute, e->Copy() );
-                    }
-                } else {
-                    dprintf( D_NEVER, "illegal type %d for attribute %s\n", (int)v.GetType(), attribute.c_str() );
-
-                    std::string buffer;
-                    classad::ClassAdUnParser caup;
-                    caup.Unparse(buffer, e);
-                    dprintf( D_NEVER, "expression was '%s'\n", buffer.c_str() );
-                }
-        } else {
-            dprintf( D_NEVER, "failed to evaluate attribute %s\n", attribute.c_str() );
+        std::vector<ExprTree *> exprs;
+        list->GetComponents(exprs);
 
-            std::string buffer;
-            classad::ClassAdUnParser caup;
-            caup.Unparse(buffer, e);
-            dprintf( D_NEVER, "expression was '%s'\n", buffer.c_str() );
+        std::vector<ExprTree *> filteredExprs;
+        for( ExprTree * expr : exprs ) {
```

Comment:
can this be `for (const ExprTree *expr : exprs) {`

Reply:
Not immediately; `isClassad()` is both capitalized wrong and not marked `const`.  I'll try changing the latter and seeing what happens...

Reply:
... looks like it Just Works(TM).  Changed.